### PR TITLE
Fix Testcase renaming on disk

### PR DESCRIPTION
**Pull request by SpaceWhite**
rel : https://github.com/AFLplusplus/LibAFL/pull/1092#issuecomment-1481551747

Sorry for very long wait.

After long time debugging, I realized that there is wrong with the lock file deletion.
Instead of passing the filename as parameter, it should pass the file name with correct path.

@tokatoka

### CODE REVIEW

File: fuzzers/fuzzbench_qemu/Makefile.toml [hunk 1]
```diff
@@ -79,7 +79,7 @@ windows_alias = "unsupported"
 script_runner = "@shell"
 script='''
 rm -rf libafl_unix_shmem_server || true
-timeout 10s cargo run --release -- --libafl-in ../libfuzzer_libpng/corpus --libafl-out ./out ./${FUZZER_NAME} >fuzz_stdout.log || true
+timeout 30s bash -c 'cargo run --release -- --libafl-in ../libfuzzer_libpng/corpus --libafl-out ./out ./${FUZZER_NAME}' > fuzz_stdout.log || true
```
Comment (tokatoka) [inline fuzzers/fuzzbench_qemu/Makefile.toml:82]: I found qemu fuzzers doesn't work on GitHub runners. (perhaps OOM?)
the fuzzers never run.
thus I removed the tests in #1179 

so can you revert this change on this file?

Comment (SpaceWhite) [reply fuzzers/fuzzbench_qemu/Makefile.toml]: done

**Comment by tokatoka** [comment]
can you cargo fmt? thanks

File: libafl/src/corpus/testcase.rs [hunk 2]
```diff
@@ -144,8 +146,53 @@ where
 
     /// Set the filename
     #[inline]
-    pub fn set_filename(&mut self, filename: String) {
+    #[cfg(feature = "std")]
+    pub fn set_filename(&mut self, filename: String) -> Result<(), Error> {
+        use std::fs::OpenOptions;
+
+        if self.filename.is_some() {
+            let f = self.filename.clone().unwrap();
```
Comment (domenukk) [inline libafl/src/corpus/testcase.rs:154]: I think it's okay to `take()` the old `filename` here, then again we don't need an extra allocation.
Also, it's a good idea to first check if the new filename is different from the old one, else do nothing :)

**Review by domenukk — CHANGES_REQUESTED** [review]
Some small comments, looks good otherwise! Thanks :)

**Comment by domenukk** [comment]
Could try these changes:

```diff
diff --git a/libafl/src/corpus/testcase.rs b/libafl/src/corpus/testcase.rs
--- libafl/src/corpus/testcase.rs
+++ libafl/src/corpus/testcase.rs
@@ -150,15 +150,15 @@
     pub fn set_filename(&mut self, filename: String) -> Result<(), Error> {
         use std::fs::OpenOptions;
 
         if self.filename.is_some() {
-            let f = self.filename.clone().unwrap();
-            let old_filename = f.as_str();
+            let old_filename = self.filename.take();
 
-            let new_filename = filename.as_str();
+            let new_filename = filename;
 
             // Do operations below when new filename is specified
-            if old_filename.eq(new_filename) {
+            if old_filename == new_filename {
+                self.filename = Some(old_filename);
                 return Ok(());
             }
 
             let new_lock_filename = format!(".{new_filename}.lafl_lock");
@@ -169,8 +169,9 @@
                 .write(true)
                 .open(&new_lock_filename)
                 .is_err()
             {
+                self.filename = Some(old_filename);
                 return Err(Error::illegal_state(
                     "unable to create lock file for new testcase",
                 ));
             }

```

and 

```diff
diff --git a/libafl/src/corpus/inmemory_ondisk.rs b/libafl/src/corpus/inmemory_ondisk.rs
--- libafl/src/corpus/inmemory_ondisk.rs
+++ libafl/src/corpus/inmemory_ondisk.rs
@@ -214,34 +214,34 @@
 
     fn save_testcase(&self, testcase: &mut Testcase<I>, idx: CorpusId) -> Result<(), Error> {
         if testcase.filename().is_none() {
             // TODO walk entry metadata to ask for pieces of filename (e.g. :havoc in AFL)
-            let file_orig = testcase.input().as_ref().unwrap().generate_name(idx.0);
-            let mut file = file_orig.clone();
+            let filename_orig = testcase.input().as_ref().unwrap().generate_name(idx.0);
+            let mut filename = filename_orig.clone();
 
             let mut ctr = 2;
-            let filename = loop {
+            let (filename, lockfile) = loop {
                 let lockfile = format!(".{file}.lafl_lock");
+                let lockfile = self.dir_path.join(lockfile);
 
                 if OpenOptions::new()
                     .write(true)
                     .create_new(true)
-                    .open(self.dir_path.join(lockfile))
+                    .open(lockfile)
                     .is_ok()
                 {
-                    break file;
+                    break (filename, lockfile);
                 }
 
-                file = format!("{file_orig}-{ctr}");
+                filename = format!("{file_orig}-{ctr}");
                 ctr += 1;
             };
 
-            let file_path = self.dir_path.join(filename.clone());
-            let filename_str = file_path.to_str().expect("Invalid Path");
+            let file_path = self.dir_path.join(&filename);
+            let filename_str = file_path.to_str()?;
             testcase.set_filename(filename_str.into())?;
 
-            let lock_file_path = self.dir_path.join(format!(".{filename}.lafl_lock"));
-            fs::remove_file(lock_file_path)?;
+            fs::remove_file(lockfile)?;
         };
         if self.meta_format.is_some() {
             let mut filename = PathBuf::from(testcase.filename().as_ref().unwrap());
             filename.set_file_name(format!(

```

**Comment by domenukk** [comment]
Will try the fixes myself, merging for now, thanks! :)

**Review by domenukk — APPROVED** [review]
